### PR TITLE
Sofias Library, w.3 Web Development Bootcamp Spring 2024

### PR DESCRIPTION
**Pull request by sofia32057**
## Netlify link
[sofias-library.netlify.app](https://sofias-library.netlify.app/)

### CODE REVIEW

**Comment by Yifan-858** [comment]
Hey Sofia, great job! 👏

The codebase is very structured and consistent flow which enhances the readability and makes it easier for other developers to navigate through the code. Good effort with implementing the search bar by using `asyc` and the "to top" button. Very inspiring!

Kudos on styling the website according to the Figma design and adding your own touch. It gives a nice and aligned visual style to the page. It looks neat with a foldable filter&sorting&searching menu. 

While reviewing the code, I've identified a few minor points. Here are my observations and suggestions for improvement:

- I've noticed when the mouse hovers over the book item, the size of the book item changes a bit. It seems to be caused by the addition of pixels in the border style. My approach is adding a transparent border style in the original book item styling, so as to make the transition smoother.
```
.book-item {
  border: transparent 2px solid;
}
```

- The button for expanding the foldable menu is labeled "Options". The function can be clearer for the users by using a more descriptive label such as "Filters & Sort". 

- The getRandomBook code doesn't check if the next random book is the same as the current one. Sometimes, the user gets the same book which makes it feel like the button doesn't work. I added a do/while loop for checking if the`Math.random()` gives the same number. 
```
const getRandomBook = () => {
  let i;
  let previousNum = -1;
  do {
    i = Math.floor(Math.random() * books.length);
  } while (i === previousNum); // prevent getting the same number
  previousNum = i; //update previousNum
  const randomBook = [books[i]];
  bookCards(randomBook);
};
```
- I also struggled with making filters and sorting interact with each other. I couldn't get multiple filters to work at the same time. but I managed to apply sorts to different filters. My approach is defining applySorting and checkSorted functions:
```
let filteredBooks = []; //will be updated by filter functions
let sortingFunction = null; // will be updated by sorting cases & checkSorted function
let previousSort = null; // will be updated in checkSorted function

const applySorting = () => {
  let activeList = filteredBooks.length === 0 ? books : filteredBooks; 
//if filteredBooks.length === 0, then there is no filter applied. display the original books array. 
//if there is value in filteredBooks, then we pass on the filtered books array to next step.

  if (sortingFunction) { //check if there is any sorting applied
    activeList = sortingFunction(activeList);
  }

  getBooks(activeList);
};

const checkSorted = (sortFunc) => {
  if (previousSort === sortFunc) {
    //Deselect sorting
    sortingFunction = null;
  } else {
    //Select sorting
    sortingFunction = sortFunc;
  }

  applySorting();
  previousSort = sortingFunction;
};

//create sort functions
const sortByAuthor = (arr) => {
   //logic for sorting
  return sortedResultt;
};

//add event listener
authorBtn.addEventListener("click", () => {
  checkSorted(sortByAuthor);
});

filterBtn.addEventListener("click", (event) => {
        filteredBooks = filterFunction();
        applySorting();
 });

     
```
I believe there is a big space for improvement. But this logic is what I can think of for now.😂 I hope it provides you with some kind of input. I would be happy if you could get some new filters and sorts solutions. 

All in all, you are doing awesome! Looking forward to more discussions in the future!

**Comment by sofia32057** [comment]
Thank you for the comment, @Yifan-858! Really interesting to see your approach. I will for sure implement a check for the randomizer - that's a really good point!

Regarding the filter/sorting I had forgotten to update the readme - the feature is done and implemented already. Both combination of sorting/filter and multiple filters. Doesn't it work for you? But your solution is definitely a good way to go :)

File: .gitignore
Comment (NellaMori) [inline .gitignore:1]: amazing 🔥

File: script.js [hunk 1]
```diff
@@ -1,406 +1,486 @@
...
+// Reset
+const reset = () => {
+  activeList = books;
+  getBooks(activeList);
+  activeFilter = "";
+  activeSort = "";
+};
+
+// Handle search
+async function getSearchResults(searchterm) {
```
Comment (NellaMori) [inline script.js]: In this code there is no asynchronous functionality, so there is no need for the async keyword in any function. All functions operate synchronously and do not involve any asynchronous operations like fetching data from an external source or performing time-consuming tasks that would require the use of asynchronous functions. You can rename you function like so:

```
const getSearchResults = (searchTerm) => {
  ...
}
```

**Review by NellaMori — CHANGES_REQUESTED** [review]
Hi Sofia, well done finishing your Library app 📚 Everything works pretty well, apart from the filter functionality. You see that right now, once a filter by genre is selected, the result is not always correct and accurate. Take a look at this bug 🐛 and test the app to make sure it works in every part. Great job with responsiveness and copying the design, one feedback about UX and UI would be to avoid elements moving when clicking on the expanding arrow to see the description. You see that the div expands for every book, even if you are selecting only one. There are different workarounds for this issue, I recommend for now to leave the description visible as you see in the Figma design (in that case with the list of ingredients). I left some comments in the code, please update the app to fix the bug and you're good to go 🚀 thanks ⭐

**Review by NellaMori — APPROVED** [review]

**Pull request closed without merging**